Rewrite Automation section after w3c/sensors#470

JuhaVainio opened the pull request:
The Automation section in the Generic Sensor API specification was rewritten and several terms and concepts have changed.

This commit adapts the Proximity Sensor spec to the changes:

- Remove references to "mock sensor type", "mock sensor reading values" and the "MockSensorType" enum.
- Define an entry in the per-type virtual sensor metadata map whose key is what used to be the "proximity" entry in MockSensorType and an appropriate virtual sensor metadata entry.

This is enough to integrate properly with the Generic Sensor spec and allow Proximity virtual sensors to be created and used.

Fixes #56.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/JuhaVainio/proximity/pull/57.html" title="Last updated on Oct 26, 2023, 7:46 AM UTC (d50353b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/proximity/57/67c242f...JuhaVainio:d50353b.html" title="Last updated on Oct 26, 2023, 7:46 AM UTC (d50353b)">Diff</a>

JuhaVainio committed:
Rewrite Automation section after w3c/sensors#470

The Automation section in the Generic Sensor API specification was rewritten and several terms and concepts have changed.

This commit adapts the Proximity Sensor spec to the changes:

Remove references to "mock sensor type", "mock sensor reading values" and the "MockSensorType" enum.
Define an entry in the per-type virtual sensor metadata map whose key is what used to be the "proximity" entry in MockSensorType and an appropriate virtual sensor metadata entry.
This is enough to integrate properly with the Generic Sensor spec and allow Proximity virtual sensors to be created and used.

Fixes #56.

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
+: [=map/key=]
+:: "`proximity`"
+: [=map/value=]
+:: A [=virtual sensor metadata=] whose [=virtual sensor metadata/virtual sensor type=] is [=Proximity Sensor=] and [=reading parsing algorithm=] is [=proximity reading parsing algorithm=].
```
```suggestion
:: A [=virtual sensor metadata=] whose [=virtual sensor metadata/virtual sensor type=] is [=Proximity Sensor=] and [=reading parsing algorithm=] is the [=proximity reading parsing algorithm=].
```

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
-  };
-</pre>
+<h3 dfn>Proximity reading parsing algorithm</h3>
+<div algorithm="proximity reading parsing algorithm">
```
You need to close this tag.

JuhaVainio replied:
Closed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
+
+    1.  Let |reading| be a new [=sensor reading=].
+    1.  Let |value| be a new [=sensor reading=].
+    1. Let |key| be the [=/list=] « "`distance`", "`max`", "`near`" ».
```
```suggestion
    1.  Let |key| be the [=/list=] « "`distance`", "`max`", "`near`" ».
```

Furthermore: this is a list, so just like you'd use a plural in code, so should you here (i.e. call it `keys`, not `key`).

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
+    1.  Let |reading| be a new [=sensor reading=].
+    1.  Let |value| be a new [=sensor reading=].
+    1. Let |key| be the [=/list=] « "`distance`", "`max`", "`near`" ».
+    1.  [=map/For each=] |key| → let |value| be the result of [=parse single-value number reading=] with |parameters| and |key|.
```
- This is not the right format. `key` is a list, not a map, see https://infra.spec.whatwg.org/#list-iterate.
- `|value|` should be declared in a separate substep.
- [`near`](https://w3c.github.io/proximity/#dom-proximitysensor-near) is a boolean, so you cannot use "parse single-value number reading" to parse it. You'll need some custom steps to parse this value. For example:

```diff
diff --git a/index.bs b/index.bs
index 4c7a0a5..c887f88 100644
--- a/index.bs
+++ b/index.bs
@@ -43,6 +43,9 @@ urlPrefix: https://w3c.github.io/sensors; spec: GENERIC-SENSOR
     text: fingerprinting; url: device-fingerprinting
     text: user identifying; url: user-identifying
     text: mitigation strategies; url: mitigation-strategies
+urlPrefix: https://w3c.github.io/webdriver/; spec: WEBDRIVER2
+  type: dfn
+    text: get a property; url: dfn-getting-properties
 </pre>
 <pre class="link-defaults">
 spec:infra;
```

and then

```
- Let |near| be the result of invoking [=get a property=] from |parameters| with "near".
- If |near|'s type is not Boolean, return **undefined**.
- Set reading["near"] to |near|.
```

JuhaVainio replied:
Hopefully fixed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
+    1.  Let |value| be a new [=sensor reading=].
+    1. Let |key| be the [=/list=] « "`distance`", "`max`", "`near`" ».
+    1.  [=map/For each=] |key| → let |value| be the result of [=parse single-value number reading=] with |parameters| and |key|.
+        1.  If |value| is "undefined".
```
You're using the `**undefined**` in the output section of the algorithm, but `"undefined"` here. Please use one or the other consistently.

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
+    1.  [=map/For each=] |key| → let |value| be the result of [=parse single-value number reading=] with |parameters| and |key|.
+        1.  If |value| is "undefined".
+            1.  Return "undefined".
+        1.  [=map/Set=] |reading|[|key|] to |value|.
```
If this step was reached, it means `value` is a map, so this assignment is wrong. You need to use `|value|[|key|]` here.

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented on index.bs:
```diff
@@ -199,23 +196,30 @@ Abstract Operations {#abstract-operations}
+    :: A [=sensor reading=] or **undefined**
+
+    1.  Let |reading| be a new [=sensor reading=].
+    1.  Let |value| be a new [=sensor reading=].
```
This step is not needed.

JuhaVainio replied:
Removed in https://github.com/w3c/proximity/pull/57/commits/28ddd12525270a98cd6f42f37ec1345a14eb9994

rakuco commented:
> Remove references to "mock sensor type", "mock sensor reading values" and the "MockSensorType" enum. Define an entry in the per-type virtual sensor metadata map whose key is what used to be the "proximity" entry in MockSensorType and an appropriate virtual sensor metadata entry. This is enough to integrate properly with the Generic Sensor spec and allow Proximity virtual sensors to be created and used.

This should be formatted as a list here and in the commit message.

JuhaVainio committed:
Change handling of key from map to list. Use "get a property" for "near"

Typo fixes.

Changed handling of "key" from map to list.

Changed "key" to "keys".

Use "get a property" operation for "near".

JuhaVainio commented:
> > Remove references to "mock sensor type", "mock sensor reading values" and the "MockSensorType" enum. Define an entry in the per-type virtual sensor metadata map whose key is what used to be the "proximity" entry in MockSensorType and an appropriate virtual sensor metadata entry. This is enough to integrate properly with the Generic Sensor spec and allow Proximity virtual sensors to be created and used.
> 
> This should be formatted as a list here and in the commit message.

Done

rakuco commented on index.bs:
```diff
@@ -199,23 +199,35 @@ Abstract Operations {#abstract-operations}
+
+    1.  Let |reading| be a new [=sensor reading=].
+    1.  Let |keys| be the [=/list=] « "`distance`", "`max`" ».
+    1.  [=map/For each=] |keys| of [=/list=]
```
* "list" in https://infra.spec.whatwg.org/#list-iterate was the name of the list, not the type.
* You are still referring to map's "for each", not list's.
* Think of each like a "for each" that iterates through a list or a vector: you need the vector and the item that will receive each item in the vector.

JuhaVainio replied:
I think

```
- "list" in https://infra.spec.whatwg.org/#list-iterate was the name of the list, not the type.
- You are still referring to map's "for each", not list's.
```

are fixed in https://github.com/w3c/proximity/pull/57/commits/8eb1313cf653d65668c523d4b9e9831356f61e73
But doubt that 

```
Think of each like a "for each" that iterates through a list or a vector: you need the vector and the item that will receive each item in the vector.
```

is... Let's see

rakuco commented on index.bs:
```diff
@@ -199,23 +199,35 @@ Abstract Operations {#abstract-operations}
+    1.  Let |near| be the result of invoking [=get a property=] from |parameters| with "near".
+        1. If |near|'s type is not Boolean
+            1.  Return **undefined**.
+        1.  [=map/Set=] reading["near"] to |near|.
```
```suggestion
        1.  [=map/Set=] |reading|["near"] to |near|.
```

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/8eb1313cf653d65668c523d4b9e9831356f61e73

JuhaVainio committed:
Change list format from map to list

Keys was incorrectly defined as map and not as list.

Defined index for keys which is passed to "parse single-value number reading".

Typo fix in variable.

rakuco commented on index.bs:
```diff
@@ -199,23 +199,35 @@ Abstract Operations {#abstract-operations}
+    1.  Let |reading| be a new [=sensor reading=].
+    1.  Let |keys| be the [=/list=] « "`distance`", "`max`" ».
+    1.  [=list/For each=] |key| of |keys|
+        1.  Let |value| be the result of [=parse single-value number reading=] with |parameters| and |key|.
```
```suggestion
        1.  Let |value| be the result of invoking [=parse single-value number reading=] with |parameters| and |key|.
```

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/0e039e64e6b1746dc5e34e2cf010be713f8a3e98

rakuco commented on index.bs:
```diff
@@ -199,23 +199,35 @@ Abstract Operations {#abstract-operations}
+            1.  If |value| is **undefined**.
+                1.  Return **undefined**.
+        1.  [=map/Set=] |reading|[|keys|] to |value|[|keys|].
+    1.  Let |near| be the result of invoking [=get a property=] from |parameters| with "near".
```
```suggestion
    1.  Let |near| be the result of invoking [=get a property=] from |parameters| with "`near`".
```

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/0e039e64e6b1746dc5e34e2cf010be713f8a3e98
Could you double check the correctness of line 224.

rakuco replied:
> Could you double check the correctness of line 224.

Good catch, I've added a suggested fix to the line.

rakuco reviewed (approved):
lgtm with a couple of nits

JuhaVainio committed:
Fix style of parameter

Added backticks to "near" when it represents text.

rakuco commented on index.bs:
```diff
@@ -199,23 +199,35 @@ Abstract Operations {#abstract-operations}
+        1.  Let |value| be the result of invoking [=parse single-value number reading=] with |parameters| and |key|.
+            1.  If |value| is **undefined**.
+                1.  Return **undefined**.
+        1.  [=map/Set=] |reading|[|keys|] to |value|[|keys|].
```
```suggestion
        1.  [=map/Set=] |reading|[|key|] to |value|[|key|].
```

JuhaVainio replied:
Fixed in https://github.com/w3c/proximity/pull/57/commits/d50353b020aeaa0e927eedea487247e6ecf6e241

JuhaVainio committed:
Fix the wrong variable name (keys to key)

rakuco reviewed (approved):
still lgtm
@anssiko can you merge this PR?

anssiko reviewed (approved):
Thanks @JuhaVainio and @rakuco!
